### PR TITLE
Add dynamic artifacts naming, documentation and tests

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I implemented the dynamic artifacts naming for output artifacts to support the following use cases:
- Static naming (as it was before)
- Dynamic via placeholder - `{time}` and `{date}` are supported out of the box + any custom placeholder can be passed via `substitutions`

```python
@step(substitutions={"my_time":"value"})
def ret_42()->Annotated[float,"dummy_{my_time}"]:
    return time.time()
    
@pipeline
def my_pipeline(substitutions={"my_time":"some_pipe_defenition"}):
    ret_42()
    ret_42(substitutions={"my_time":"another_value"})()
```

The same applies to `ArtifactConfig(name=...)` pattern.

There are breaking changes for frontend since now the user can see something like this:
image

The actual substitutes of the step can be acquired from the `PipelineRunResponse`, but on frontend the evaluation of the templated names from `DeploymentResponse` should happen anyways to map those "pre-defined" outputs once they are created on backend.

**DO NOT MERGE BEFORE FE CHANGES ARE READY**

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: docs/book/how-to/data-artifact-management/handle-data-artifacts/artifacts-naming.md
Comment (htahir1) [inline docs/book/how-to/data-artifact-management/handle-data-artifacts/artifacts-naming.md:1]: Needs to be in toc.md

File: docs/book/how-to/data-artifact-management/handle-data-artifacts/artifacts-naming.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+---
+description: Understand how you can name your ZenML artifacts.
+---
+
+# How Artifact Naming works in ZenML
```
Comment (htahir1) [inline docs/book/how-to/data-artifact-management/handle-data-artifacts/artifacts-naming.md:5]: Do you want to remove these pages then? https://docs.zenml.io/how-to/pipeline-development/build-pipelines/dynamically-assign-artifact-names

**Review by htahir1 — CHANGES_REQUESTED** [review]

File: src/zenml/artifacts/artifact_config.py [hunk 2]
```diff
@@ -83,3 +91,35 @@ def _remove_old_attributes(cls, data: Dict[str, Any]) -> Dict[str, Any]:
             )
 
         return data
+
+    @model_validator(mode="after")
+    def artifact_config_after_validator(self) -> "ArtifactConfig":
```
Comment (schustmi) [inline src/zenml/artifacts/artifact_config.py:96]: I'm wondering why this is happening as an after validator. This means if I run the same pipeline twice in one python script, it will evaluate just once and therefore reuse the same name?

If that is indeed the case, is this intended? IMO I'd expect the callable to be evaluated each time a pipeline is being run, and we could easily do that by just doing the evaluation in the `_evaluated_name` property?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/steps/utils.py [hunk 3]
```diff
@@ -249,6 +281,10 @@ def get_artifact_config_from_annotation_metadata(
             if artifact_config is not None:
                 raise ValueError(error_message)
             artifact_config = metadata_instance
+        elif callable(metadata_instance):
+            if output_name is not None:
+                raise ValueError(error_message)
+            output_name = metadata_instance
```
Comment (schustmi) [inline src/zenml/steps/utils.py:287]: In line `300` below I think might be another issue: What if someone passes a callable and an artifact config without name. Seems like it then sets the callable on the artifact config, but it does not get evaluated (because the validator already ran)

File: src/zenml/artifacts/artifact_config.py [hunk 4]
```diff
@@ -83,3 +91,35 @@ def _remove_old_attributes(cls, data: Dict[str, Any]) -> Dict[str, Any]:
...
+    def _evaluated_name(self) -> Optional[str]:
+        """Evaluated name of the artifact.
+
+        Returns:
+            The evaluated name of the artifact.
+
+        Raises:
+            RuntimeError: If the name is still a callable.
+        """
+        if callable(self.name):
```
Comment (schustmi) [inline src/zenml/artifacts/artifact_config.py:121]: Do we want to reuse the date/time from the pipeline run here, same as we do with model version names? If it's even possible to pass this here, I'm not sure

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/artifacts/artifact_config.py (hunk 2 above)
Comment (avishniakov) [reply src/zenml/artifacts/artifact_config.py:96]: This is a valid concern, I acknowledge it. I started exploring this and it turned out to be quite a complex issue to solve (though I would still do it). It hits exactly the same limitations with step configs being saved and reused from the deployment, so I would need to implement something to break this circle here already, not you as part of Lazy Loaders rework.

File: src/zenml/artifacts/artifact_config.py (hunk 4 above)
Comment (avishniakov) [reply src/zenml/artifacts/artifact_config.py:121]: Might be interesting. Maybe this can be set on a context level instead of pulling fresh each time.

File: src/zenml/artifacts/artifact_config.py [hunk 5]
```diff
@@ -45,7 +47,10 @@ def my_step() -> Annotated[
     ```
 
     Attributes:
-        name: The name of the artifact.
+        name: The name of the artifact:
+            - static string e.g. "name"
+            - dynamic callable e.g. lambda: "name"+str(42)
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/artifacts/artifact_config.py:52]: Thanks, fixed this part

File: src/zenml/artifacts/artifact_config.py (hunk 4 above)
Comment (avishniakov) [reply src/zenml/artifacts/artifact_config.py]: I didn't implement this, but I can give it a shot if the PR is conceptually fine with you.

File: src/zenml/models/v2/core/artifact.py [hunk 6]
```diff
@@ -42,6 +42,11 @@ class ArtifactRequest(BaseRequest):
         title="Name of the artifact.",
         max_length=STR_FIELD_MAX_LENGTH,
     )
+    original_name: Optional[str] = Field(
```
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/models/v2/core/artifact.py:45]: What happens if I create two artifacts which have the same name, but different original names? I don't think this would work as expected, no?

Shouldn't the mapping of step runs to its output artifact happen in the `StepRunOutputArtifactSchema`, which is already storing a name?

In what other occasions do we need to map information stored in the deployment to these artifacts?

[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/models/v2/core/artifact.py]: You are right that it should not belong to the `Artifact` because of the duality you mentioned, but I also don't feel it belongs to the `StepRunOutputArtifactSchema`, rather to `ArtifactVersionSchema`, thus I made changes to add it there and added extra test to cover that edge case.
I don't know about any other use cases, but this is surely the case for the new DAG with all steps/outputs shown at once.

[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/models/v2/core/artifact.py]: What is the reasoning the makes sense on the artifact version though? Isn't this something that explicitly happens on step outputs? When someone manually uploads an artifact version, there would never be an `original_name`, no? Somehow it just feels unintuitive for me right now, ideally we wouldn't even need to store this IMO. Is there any way we can get around storing this at all?

[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/models/v2/core/artifact.py]: IMO, the reasoning is that this information was the driver for the Artifact Version to receive the name if we can find it in the Artifact entity. However, it can come from various sources (static or dynamic). I don't see why this should relate to the Step Output directly and overcomplicate the schema relations while filling that info into other entities. If the artifact version has no original name, e.g. for manual save or for static names - this field is just not populated in DB and only backfilled in runtime with the artifact name itself. So, with the current implementation, it would be close to no impact if the users are still relying on static naming, or with a small overhead for dynamic names to store their original source of naming.

[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/models/v2/core/artifact.py]: Regarding the second question: I doubt that there is a good way to avoid storing something if we keep the current DAG filling logic.
What I can think of as a solution could be:
- On run start we show the outputs from the Deployment only (since no info in the Run)
- Once the step was finished ly replace that part with the data from there

I don't feel that there is any harm in doing so, but I would love to hear from @Cahllagerfeld and @znegrin on that matter.

**Comment by avishniakov** [comment]
@Cahllagerfeld let me know what would be the most efficient way to prepare the data for you, as now substitutes are quite distributed:

There are breaking changes for frontend since now the user can see something like this:
image

The current way of thinking to approach this is the following:
- We know the templated names from the deployment (e.g. `my_name_{time}_{ck_pipe}_{ck_step`)
- We know the pipeline level substitutes from the `PipelineRunResponse` (`ck_pipe` is here)
- We know the step level substitutes from the `StepRunResponse` (`ck_step` is here + overrides for `ck_pipe` may be here, same applies to overrides of the `time` and `date`)
- We know `date` and `time` substitutes from the `start_time` of the `PipelineRunResponse` (`time` and `date` can be derived from it)

It seems logical to populate `StepRunResponse` properly on the backend with all needed substitutes, so we do not compute on the FE. Would this do? Any contras @schustmi @Cahllagerfeld ?

File: src/zenml/steps/utils.py [hunk 7]
```diff
@@ -112,6 +124,20 @@ def parse_return_type_annotations(
     Returns:
         - A dictionary mapping output names to their output signatures.
     """
+
+    def _define_output_name(
+        artifact_config: Optional["ArtifactConfig"],
+        i: int,
+        original_output_names: Optional[List[str]] = original_output_names,
+    ) -> Optional[str]:
+        output_name: Optional[str] = None
```
Comment (schustmi) [inline src/zenml/steps/utils.py:133]: Is this still necessary? `artifact_config.name` will always be the `original name`, correct?

File: src/zenml/orchestrators/step_runner.py [hunk 8]
```diff
@@ -149,9 +150,30 @@ def run(
...
             output_annotations = parse_return_type_annotations(
-                func=step_instance.entrypoint
+                func=step_instance.entrypoint,
+                original_output_names=list(output_materializers.keys()),
             )
 
+            substitution_dict = pipeline_run.config.full_substitutions(
+                pipeline_run.start_time or datetime.utcnow()
+            )
+            substitution_dict.update(step_run.config.substitutions or {})
```
Comment (schustmi) [inline src/zenml/orchestrators/step_runner.py:160]: Can be simplified with compiler changes

File: src/zenml/pipelines/pipeline_definition.py [hunk 9]
```diff
@@ -656,7 +662,8 @@ def _create_deployment(
                 schedule_name = schedule.name
             else:
                 schedule_name = format_name_template(
-                    deployment.run_name_template
+                    deployment.run_name_template,
+                    **(deployment.pipeline_configuration.substitutions or {}),
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/pipelines/pipeline_definition.py:666]: I played with this back in day, but it might be causing the lost of the values, if nothing is passed. E.g. substitutions are `{'a':'b'}` and empty update will make it `{}`. Thus kept in sync with other configs with None defaults

[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/pipelines/pipeline_definition.py:666]: If I look at `PipelineConfigurationUpdate` though, both `extra` and `settings` are non-optional though? And I'm sure we don't lose any values in empty updates

File: src/zenml/utils/string_utils.py [hunk 10]
```diff
@@ -161,16 +162,39 @@ def format_name_template(
...
-    kwargs["date"] = kwargs.get(
-        "date",
-        datetime.datetime.now(datetime.timezone.utc).strftime("%Y_%m_%d"),
-    )
-    kwargs["time"] = kwargs.get(
-        "time",
-        datetime.datetime.now(datetime.timezone.utc).strftime("%H_%M_%S_%f"),
-    )
-    return name_template.format(**kwargs)
+    pattern = r"\{[^}]+\}"
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/utils/string_utils.py:169]: Patter is an overkill indeed, but a safety gate to not import `get_step_context` all the time - I would keep, since it might be inefficient on frequent calls otherwise

File: src/zenml/pipelines/pipeline_definition.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/pipelines/pipeline_definition.py]: did so

File: src/zenml/config/compiler.py [hunk 11]
```diff
@@ -99,7 +99,10 @@ def compile(
 
         self._apply_stack_default_settings(pipeline=pipeline, stack=stack)
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/config/compiler.py:100]: Does `test_that_overrides_work_as_expected` cover what you expect from it?

File: src/zenml/models/v2/core/pipeline_run.py [hunk 12]
```diff
@@ -447,6 +450,17 @@ def model_version_id(self) -> Optional[UUID]:
...
         return self.get_body().model_version_id
 
+    @property
+    def substitutions(self) -> Dict[str, Dict[str, str]]:
+        """The `substitutions` property.
+
+        Returns:
+            the value of the property.
+        """
+        return self.get_metadata().config._get_full_substitutions(
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/models/v2/core/pipeline_run.py:460]: Thanks, it is not needed at all, IMO. I will remove it

File: src/zenml/orchestrators/step_runner.py [hunk 13]
```diff
@@ -152,6 +152,22 @@ def run(
                 func=step_instance.entrypoint
             )
 
+            for k, v in list(output_annotations.items()):
+                if v.artifact_config:
+                    _evaluated_name = v.artifact_config._evaluated_name(
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/orchestrators/step_runner.py:157]: Yeah, but I replaced it all the way down just to preserve the order of the keys and make this process a bit simpler. I can also p[reserve the keys order by other means, if you feel that it is important not to replace static entries.

[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/orchestrators/step_runner.py:157]: Mhm not sure I get what you mean by preserving the order here. Doesn't the current code remove all items for which `v.artifact_config.name is not None` and the others stay in the dict, so the order gets modified?

**Review by schustmi — CHANGES_REQUESTED** [review]
One more general question: From what I can see, you now removed the `PipelineRunResponse.metadata.substitutions`, which previously contained the substitutions for all the steps.

If I remember correctly that was so the frontend has these substitutions available in the DAG visualizer without querying hydrated versions of all steps. Did you think of an alternative solution to make the matching work in the frontend?

File: src/zenml/orchestrators/step_runner.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/orchestrators/step_runner.py:157]: Since 3.7 the dict keys order is guaranteed and we were relying on that. 
Let me try to explain by example:
- say your output is `["a1","a_{time}","a2"]` - dynamic is sitting between static ones
- you return `return 1,42,2` and expect `a1=1, a_10:40:20=42, a2=2`
- in the loop you are referring to now we do:
  - evaluate `a1` to `a1`, pop and insert - the dict keys are `a_{time},a2,a1` now
  - `a_{time}` to `a_10:40:20` - the dict keys are `a2,a1,a_10:40:20`
  - `a2` is `a2` and the dict keys become `a1,a_10:40:20,a2`
- All your artifacts get desired values after the step finishes

Now the same, but without poping and reassigning the static ones:
- Only `a_{time}` gets evaluated to something new, so the dict keys would be `a1,a2,a_10:40:20`
- Values assigned to `a2` and `a_10:40:20` would be mixed with one another

For that reason I always pop and reassign. Does it make sense?

[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/orchestrators/step_runner.py:157]: Yes absolutely, but what if `v.artifact_config is None`, or `evaluated_name is None`? Shouldn't this be happening in all cases to make sure the order is preserved?

**Comment by avishniakov** [comment]
> One more general question: From what I can see, you now removed the `PipelineRunResponse.metadata.substitutions`, which previously contained the substitutions for all the steps.
> 
> If I remember correctly that was so the frontend has these substitutions available in the DAG visualizer without querying hydrated versions of all steps. Did you think of an alternative solution to make the matching work in the frontend?

You are right here - I removed it ignoring the fact that steps ar not hydrated there. I brought it back as `steps_substitutions`

File: src/zenml/orchestrators/step_runner.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/orchestrators/step_runner.py]: Reworked this piece and extended the tests set

File: src/zenml/models/v2/core/pipeline_run.py [hunk 14]
```diff
@@ -237,6 +237,9 @@ class PipelineRunResponseMetadata(WorkspaceScopedResponseMetadata):
         default=False,
         description="Whether a template can be created from this run.",
     )
+    steps_substitutions: Dict[str, Dict[str, str]] = Field(
+        title="Substitutions used in the step runs of this pipeline run.",
```
Comment (schustmi) [inline src/zenml/models/v2/core/pipeline_run.py:241]: If we give this a default value, we might avoid some client-server mismatch errors? Do you think it messes with something if we do that?

Comment (avishniakov) [reply src/zenml/models/v2/core/pipeline_run.py:241]: Nope, shall be good to go, added the `default_factory`

**Review by schustmi — APPROVED** [review]
Sorry for the long review process, this looks great now!
